File: Recursion/lecture 12/Fraz_recursion_12.java
import java.util.*;

public class Solution {
    public static ArrayList<ArrayList<Integer>> result;

    private static void solve(ArrayList<Integer> currList, int currIndex, int currSum, int b, ArrayList<Integer> arr) {
        if (currSum == b) 
        {
            result.add(new ArrayList<Integer>(currList));
            return;
        }
        if (currIndex == arr.size()) 
        {
            return;
        }

        solve(currList, currIndex + 1, currSum, b, arr);

        int count = 0;

        while (currSum <= b) 
        {
            count++;

            currSum += arr.get(currIndex);

            currList.add(arr.get(currIndex));

            solve(currList, currIndex + 1, currSum, b, arr);
        }
        while (count > 0) 
        {
            int index = currList.size() - 1;
            currList.remove(index);
            count--;
        }
    }

    public static ArrayList<ArrayList<Integer>> combSum(ArrayList<Integer> arr, int b) 
    {

        Collections.sort(arr);
        result = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> currList = new ArrayList<Integer>();
        solve(currList, 0, 0, b, arr);
        return result;
    }
}

File: Recursion/lecture 12/Fraz_recursion_12.py
from copy import copy

def solve(result, currIndex, currSum, currList, B, ARR):

    if currSum == B:
        result.append(copy(currList))
        return

    if currIndex == len(ARR):
        return 

    solve(result, currIndex + 1, currSum, currList, B, ARR)


    count = 0

    while currSum <= B:

        count += 1

        currSum += ARR[currIndex]

        currList.append(ARR[currIndex])

        solve(result, currIndex + 1, currSum, currList, B, ARR)
    while count:
        currList.pop()
        count -= 1


def combSum(ARR, B):

    ARR.sort()
    
    result = []
    currList = []

    solve(result, 0, 0, currList, B, ARR)

    return result